FirebaseSessions/Sources/Settings.swift

```diff
@@ -32,47 +33,77 @@ class Settings: SettingsProtocol {
   private static let flagSamplingRate = "sampling_rate"
-  private static let flagSessionTimeout = "session_timeout"
+  private static let flagSessionTimeout = "session_timeout_seconds"
   private static let flagCacheDuration = "cache_duration"
-  private let cache: SettingsCacheClient
+  private static let flagSessionsCache = "app_quality"
   private let appInfo: ApplicationInfoProtocol
+  private let downloader: SettingsDownloadClient
+  private var cache: SettingsCacheClient
 
   var sessionsEnabled: Bool {
-    guard let enabled = cache.cacheContent?[Settings.flagSessionsEnabled] as? Bool else {
+    guard let enabled = sessionsCache[Settings.flagSessionsEnabled] as? Bool else {
       return true
     }
     return enabled
   }
 
   var samplingRate: Double {
-    guard let rate = cache.cacheContent?[Settings.flagSamplingRate] as? Double else {
+    guard let rate = sessionsCache[Settings.flagSamplingRate] as? Double else {
       return 1.0
     }
     return rate
   }
 
   var sessionTimeout: TimeInterval {
-    guard let timeout = cache.cacheContent?[Settings.flagSessionTimeout] as? Double else {
+    guard let timeout = sessionsCache[Settings.flagSessionTimeout] as? Double else {
       return 30 * 60
     }
     return timeout
   }
 
   private var cacheDurationSeconds: TimeInterval {
-    guard let duration = cache.cacheContent?[Settings.flagCacheDuration] as? Double else {
+    guard let duration = cache.cacheContent[Settings.flagCacheDuration] as? Double else {
       return Settings.cacheDurationSecondsDefault
     }
     return duration
   }
 
-  init(cache: SettingsCacheClient = SettingsCache(),
-       appInfo: ApplicationInfoProtocol) {
-    self.cache = cache
+  private var sessionsCache: [String: Any] {
+    return cache.cacheContent[Settings.flagSessionsCache] as? [String: Any] ?? [:]
+  }
+
+  init(appInfo: ApplicationInfoProtocol,
+       downloader: SettingsDownloadClient,
+       cache: SettingsCacheClient = SettingsCache()) {
     self.appInfo = appInfo
+    self.cache = cache
+    self.downloader = downloader
+  }
+
+  func fetchAndCacheSettings(currentTime: Date) {
+    // Only fetch if cache is expired, otherwise do nothing
+    guard isCacheExpired(currentTime: currentTime) else {
+      Logger.logDebug("[Settings] Cache is not expired, no fetch will be made.")
+      return
+    }
+
+    downloader.fetch { result in
+      switch result {
+      case let .success(dictionary):
+        self.cache.cacheContent = dictionary
+        self.cache.cacheKey = CacheKey(
+          createdAt: currentTime,
+          googleAppID: self.appInfo.appID,
+          appVersion: self.appInfo.synthesizedVersion
+        )
+      case let .failure(error):
+        Logger.logError("[Settings] Fetching newest settings failed with error: \(error)")
+      }
+    }
   }
 
-  func isCacheExpired(currentTime: Date) -> Bool {
-    guard cache.cacheContent != nil else {
+  private func isCacheExpired(currentTime: Date) -> Bool {
```

Comment:
Tried changing it but it breaks tests because this is a private method, tests needs to pass down a fake date to every method. Considering that and since this is a private method only used by `fetch` I think it's fine to keep it no-defaults.

Reply:
Sounds good. In that my only (nit) suggestion would be to name the parameter as just `time` instead of `currentTime`.

Reply:
Done